storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+        if (transition.isDefined) {
+          val nextState = transition.get.get(currentSymbol)
+          if (nextState.isDefined) {
+            next = nextState.get :: Nil
+          }
+        }
+      }
+      Some(next)
+    }
+
+    //state is final unless the original was
+    def isFinal(statesSet: List[Int]): Boolean =
+      !(statesSet.nonEmpty && finalStates.contains(statesSet.head))
+
+    Fsm.crawl(alphabet, resInitial, isFinal, follow).Reduced
+  }
+
+  /**
+    * Generate strings (lists of symbols) that this FSM accepts. Since there may
+    * be infinitely many of these we use a Stream[String] instead of constructing a
+    * static list. Strings will be sorted in order of length and then lexically.
+    * This procedure uses arbitrary amounts of memory but is very fast. There
+    * may be more efficient ways to do this, that I haven't investigated yet.
+    * You can use this in list comprehensions.
+    */
+  def Strings: Stream[String] = {
+    // Many FSMs have "dead states". Once you reach a dead state, you can no
+    // longer reach a final state. Since many strings may end up here, it's
+    // advantageous to constrain our search to live states only.
+    val livestates = states.filter(state => IsLive(state))
+
+    // We store a list of tuples. Each tuple consists of an input string and the
+    // state that this input string leads to. This means we don't have to run the
+    // state machine from the very beginning every time we want to check a new
+    // string.
+    val pending = Queue[(String, Int)]()
+    val pendingFinals = Queue[String]()
+    var needInitial = true
+
+    def nextStr: Option[String] =
+      if (pendingFinals.nonEmpty) {
+        Some(pendingFinals.dequeue())
+      } else {
+        //Initial entry (or possibly not, in which case this is a short one)
+        if (needInitial) {
+          needInitial = false
+          val cstate = this.initialState
+          if (livestates.contains(cstate)) {
+            pending += "" -> cstate
+            if (finalStates.contains(cstate)) {
+              pendingFinals += ""
+            }
+          }
+        }
+
+        while (pendingFinals.isEmpty && pending.nonEmpty) {
+          val (pendingString, pendingState) = pending.dequeue()
+
+          val transition = transitions.get(pendingState)
+          if (transition.isDefined) {
+            for ((nextSymbol, nextState) <- transition.get) {
+              val nextString = pendingString + nextSymbol
+
+              if (livestates.contains(nextState)) {
+                pending += nextString -> nextState
+                if (finalStates.contains(nextState)) {
+                  pendingFinals += nextString
+                }
+              }
+            }
+          }
+        }
+
+        if (pendingFinals.nonEmpty) {
+          Some(pendingFinals.dequeue())
+        } else {
+          None
+        }
+      }
+
+    def infStream: Stream[String] = {
```

Comment:
infStream was an obfuscating name here. The stream is finite for some FSMs but might be infinite for others (therefore I've used streams there). Decided to leave as is, but rename to "genStream" to avoid any future misunderstandings.